pkg/repository/udmrepo/kopialib/lib_repo.go

```diff
@@ -0,0 +1,587 @@
+/*
+Copyright the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kopialib
+
+import (
+	"context"
+	"os"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/kopia/kopia/repo"
+	"github.com/kopia/kopia/repo/compression"
+	"github.com/kopia/kopia/repo/content/index"
+	"github.com/kopia/kopia/repo/maintenance"
+	"github.com/kopia/kopia/repo/manifest"
+	"github.com/kopia/kopia/repo/object"
+	"github.com/kopia/kopia/snapshot/snapshotmaintenance"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/vmware-tanzu/velero/pkg/repository/udmrepo"
+	"github.com/vmware-tanzu/velero/pkg/util/logging"
+)
+
+type kopiaRepoService struct {
+	logger logrus.FieldLogger
+}
+
+type kopiaRepository struct {
+	rawRepo     repo.Repository
+	rawWriter   repo.RepositoryWriter
+	description string
+	uploaded    int64
+	openTime    time.Time
+	throttle    logThrottle
+	logger      logrus.FieldLogger
+}
+
+type kopiaMaintenance struct {
+	mode      maintenance.Mode
+	startTime time.Time
+	uploaded  int64
+	throttle  logThrottle
+	logger    logrus.FieldLogger
+}
+
+type logThrottle struct {
+	lastTime int64
+	interval time.Duration
+}
+
+type kopiaObjectReader struct {
+	rawReader object.Reader
+}
+
+type kopiaObjectWriter struct {
+	rawWriter object.Writer
+}
+
+const (
+	defaultLogInterval             = time.Duration(time.Second * 10)
+	defaultMaintainCheckPeriod     = time.Hour
+	overwriteFullMaintainInterval  = time.Duration(0)
+	overwriteQuickMaintainInterval = time.Duration(0)
+)
+
+var kopiaRepoOpen = repo.Open
+
+// NewKopiaRepoService creates an instance of BackupRepoService implemented by Kopia
+func NewKopiaRepoService(logger logrus.FieldLogger) udmrepo.BackupRepoService {
+	ks := &kopiaRepoService{
+		logger: logger,
+	}
+
+	return ks
+}
+
+func (ks *kopiaRepoService) Init(ctx context.Context, repoOption udmrepo.RepoOptions, createNew bool) error {
+	repoCtx := logging.SetupKopiaLog(ctx, ks.logger)
+
+	if createNew {
+		if err := CreateBackupRepo(repoCtx, repoOption); err != nil {
+			return err
+		}
+
+		return writeInitParameters(repoCtx, repoOption, ks.logger)
+	} else {
+		return ConnectBackupRepo(repoCtx, repoOption)
+	}
+}
+
+func (ks *kopiaRepoService) Open(ctx context.Context, repoOption udmrepo.RepoOptions) (udmrepo.BackupRepo, error) {
+	repoConfig := repoOption.ConfigFilePath
+	if repoConfig == "" {
+		return nil, errors.New("invalid config file path")
+	}
+
+	if _, err := os.Stat(repoConfig); os.IsNotExist(err) {
```

Comment:
Do we need to check other error?

Reply:
Seems not.
Here the check is only to fail earlier and give a clearer message when the config file doesn't exist.
If there is any other case that need to fail earlier, we can also do the check here, but right now, we don't have.

Later the function `openKopiaRepo` will read the config file, if any problem happens, it will return with the error. Therefore, we still have opportunity to catch the error.